### PR TITLE
fix(dpg): `ClientOptions.ServiceVersion` should be pascal case

### LINKED ISSUE #1524: Preview versions are generated with lowercase "-preview"
Since in .NET names are supposed to be PascalCase, I created Roslyn analyzer ACZ0016 but it turns out there are 17 violations - all in generated code. I could suppress these, but it'll be risky because I'll have to declare hand-written models and any future generation may conflict. Instead, I think the generator should instead incur the breaking change because, for any existing values *not* in generated code, those have already shipped (presumably) and can be exempted.

### PR DESCRIPTION
**Pull request by archerzz**
# Description

For `ApiVersion` like `2020-10-01-preview`, the generator right now generates `ServiceVersion` like `V2020_10_01_preview` which will trigger Azure SDk error AZC0016 because not all nouns in the name are upper case.
This commit will fix this issue by invoking `TextInfo.ToTitleCase` which will make sure the name is in pascal case.

Fix #1524

# Checklist

To ensure a quick review and merge, please ensure:
- [x] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [x] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by archerzz** [comment]
Since only `WebPubSub` has been GAed and [their ServiceVersion](https://github.com/Azure/azure-sdk-for-net/blob/91bdcd14941e227e1bff45413062e93e2bbb0061/sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClientOptions.ServiceVersion.cs#L35) doesn't have case problem (because it's not `preview` or `beta`), I think it's OK to just fix this problem without concerns about breaking changes.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs
Comment (pshao25) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs:13]: I'm a little bit uncomfortable involving CultureInfo. Is there any specific reason? Could we implement this ourselves since it is a simple one. CultureInfo always has unpredictable pitfalls.

Comment (archerzz) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs:13]: I think it's OK since we've specified the language locale. And we know we use it to format codes, e.g. the service version which should be a combination of alphanumeric plus `-`.

**Comment by archerzz** [comment]
I checked all impacted SDKs.
| Package                                     | Problematic ServiceVersion                                                                                                                                                                        | GA? | Comments                                                                                                                                                                                                                              |
|---------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Azure.Verticals.AgriFood.Farming            | https://github.com/Azure/azure-sdk-for-net/blob/ce583117629ae55081f0744ab601dd9ccd7404f4/sdk/agrifood/Azure.Verticals.AgriFood.Farming/src/Generated/FarmBeatsClientOptions.cs#L32                | N   |                                                                                                                                                                                                                                       |
| Azure.AI.AnomalyDetector                    | https://github.com/Azure/azure-sdk-for-net/blob/983de9cfed373188d2f5af90d9c3f5423cf763d9/sdk/anomalydetector/Azure.AI.AnomalyDetector/src/Generated/AnomalyDetectorClientOptions.cs#L32           | N   |                                                                                                                                                                                                                                       |
| Azure.Communication.NetworkTraversal        | https://github.com/Azure/azure-sdk-for-net/blob/6de9067733a50e807829d05fa3f308baf4f1a098/sdk/communication/Azure.Communication.NetworkTraversal/src/CommunicationRelayClientOptions.cs#L28        | Y   | GA with non-preview api version, see here: https://github.com/Azure/azure-sdk-for-net/blob/1f598152d3550360b04bd0fdc5fc7bf10b36dac4/sdk/communication/Azure.Communication.NetworkTraversal/src/CommunicationRelayClientOptions.cs#L17 |
| Azure.Communication.PhoneNumbers            | https://github.com/Azure/azure-sdk-for-net/blob/96214b17c5eb2f85f4ebfad951b10650e728e275/sdk/communication/Azure.Communication.PhoneNumbers/src/SipRouting/SipRoutingClientOptions.cs#L28         | Y   | GA with non-preview api version, see here: https://github.com/Azure/azure-sdk-for-net/blob/2fc19761c0ed75c9481b892d1596e81d57021d84/sdk/communication/Azure.Communication.PhoneNumbers/src/Generated/PhoneNumbersClientOptions.cs#L16 |
| Azure.IoT.DeviceUpdate                      | https://github.com/Azure/azure-sdk-for-net/blob/d17b6db0aedb4f5335553bda7bf7ebfc0d0444b8/sdk/deviceupdate/Azure.IoT.DeviceUpdate/src/Generated/DeviceUpdateClientOptions.cs#L22                   | N   |                                                                                                                                                                                                                                       |
| Azure.AI.FormRecognizer                     | https://github.com/Azure/azure-sdk-for-net/blob/30f915ece20e2dd1da6197e8501785b1f7e6c9a4/sdk/formrecognizer/Azure.AI.FormRecognizer/src/DocumentAnalysisClientOptions.cs#L26                      | Y   | GA with non-preview api version, see here: https://github.com/Azure/azure-sdk-for-net/blob/e3136705a627e057ca962baf26940e3664d22c42/sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormRecognizerClientOptions.cs#L16                 |
| Azure.MixedReality.Authentication           | https://github.com/Azure/azure-sdk-for-net/blob/8ba15b38b60e9bf3d7b687a267debec173e2ee67/sdk/mixedreality/Azure.MixedReality.Authentication/src/MixedRealityStsClientOptions.cs#L41               | Y   | problematic service version in GA version: https://github.com/Azure/azure-sdk-for-net/blob/a7dfb60ee67e9a0392b692d59d1bae9c46c7b766/sdk/mixedreality/Azure.MixedReality.Authentication/src/MixedRealityStsClientOptions.cs#L26        |
| Azure.Monitor.Ingestion                     | https://github.com/Azure/azure-sdk-for-net/blob/f4f6bf0c5f8a9b7ef94599ec48d7b28f55c58d67/sdk/monitor/Azure.Monitor.Ingestion/src/Generated/LogsIngestionClientOptions.cs#L22                      | N   |                                                                                                                                                                                                                                       |
| Azure.MixedReality.ObjectAnchors.Conversion | https://github.com/Azure/azure-sdk-for-net/blob/e7f6846812af0756d2278df605e06ebeb74c7934/sdk/personalizer/Azure.AI.Personalizer/src/Models/PersonalizerClientOptions.cs#L24                       | N   |                                                                                                                                                                                                                                       |
| Azure.AI.Personalizer                       | https://github.com/Azure/azure-sdk-for-net/blob/e7f6846812af0756d2278df605e06ebeb74c7934/sdk/personalizer/Azure.AI.Personalizer/src/Models/PersonalizerClientOptions.cs#L24                       | N   |                                                                                                                                                                                                                                       |
| Azure.Analytics.Purview.Account             | https://github.com/Azure/azure-sdk-for-net/blob/dc6d5577ec3b6cb2fd97c5316d9b21c73a0e1922/sdk/purview/Azure.Analytics.Purview.Account/src/Generated/PurviewAccountClientOptions.cs#L22             | N   |                                                                                                                                                                                                                                       |
| Azure.Analytics.Purview.Administration      | https://github.com/Azure/azure-sdk-for-net/blob/051b9691594b26f64cf95e10bee4872a6084e6b9/sdk/purview/Azure.Analytics.Purview.Administration/src/Customizations/PurviewAccountClientOptions.cs#L25 | N   |                                                                                                                                                                                                                                       |
| Azure.Analytics.Purview.Catalog             | https://github.com/Azure/azure-sdk-for-net/blob/586e3d622ea98e33a2682513c44c97aba195715d/sdk/purview/Azure.Analytics.Purview.Catalog/src/Generated/PurviewCatalogClientOptions.cs#L22             | N   |                                                                                                                                                                                                                                       |
| Azure.Analytics.Purview.Scanning            | https://github.com/Azure/azure-sdk-for-net/blob/d661cc7fde3ea82fedd022c54d1a3bfb11642339/sdk/purview/Azure.Analytics.Purview.Scanning/src/Generated/PurviewScanningServiceClientOptions.cs#L22    | N   |                                                                                                                                                                                                                                       |
| Azure.Analytics.Synapse.Artifacts           | https://github.com/Azure/azure-sdk-for-net/blob/830f96a87e3365aa7b0cf9f46b542d02a7089130/sdk/synapse/Azure.Analytics.Synapse.Artifacts/src/Generated/ArtifactsClientOptions.cs#L24                | N   |                                                                                                                                                                                                                                       |
| Azure.Analytics.Synapse.Monitoring          | https://github.com/Azure/azure-sdk-for-net/blob/c669c604ee04d6fd92abad30141a19d21df3931f/sdk/synapse/Azure.Analytics.Synapse.Monitoring/src/Generated/MonitoringClientOptions.cs#L22              | N   |

All SDKs should be fine, because:
- Either they are beta
- Or their official release doesn't carry preview `ServiceVersion`.

**Except for** `Azure.MixedReality.Authentication`, which is released with preview `ServiceVersion`. However, I think we can continue with breaking change, because it has only one `ServiceVersion` so it's very unlikely that people will refer to a specific `ServiceVersion`.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs
Comment (pshao25) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs:13]: We only specify the locale. Locale cannot specify formats now. Is there any place we could confirm `-` must be explained as a delimiter?

**Comment by lirenhe** [comment]
@AlexanderSher,  as you are also working on HLC, could you help to check the above comment?  Thanks

**Comment by lirenhe** [comment]
cc @jsquire

File: src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs
Comment (archerzz) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs:13]: It's a typo, it should be underscore `_`.

**Review by jsquire — CHANGES_REQUESTED** [review]
This constitutes a breaking change.  While I appreciate your analysis and agree that the preview libraries that you've identified are permissible, we cannot move forward with this change without explicit sign-off from the owners of the impacted packages or architect approval.

Further, this change cannot be applied to existing versions in `Azure.MixedReality.Authentication` as it would be a breaking change to a stable library without an exception from the architects.   What is your plan here?

**Comment by jsquire** [comment]
//fyi:  @m-nash, for awareness

File: src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs
Comment (archerzz) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs:43]: renamed `s` to `version`

File: samples/Azure.AI.DocumentTranslation/Generated/DocumentTranslationClientOptions.cs
Comment (archerzz) [reply samples/Azure.AI.DocumentTranslation/Generated/DocumentTranslationClientOptions.cs:22]: Good catch. Thanks.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs
Comment (archerzz) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs:13]: Thanks. I wasn't aware of `CultureInfo.InvariantCulture.TextInfo`.

File: samples/Azure.AI.DocumentTranslation/Generated/DocumentTranslationClientOptions.cs
Comment (pshao25) [reply samples/Azure.AI.DocumentTranslation/Generated/DocumentTranslationClientOptions.cs:22]: Just curious. Is the version a free text? How could we gaurantee is follows the specific format now?

FYI, we will have such linter for data plane
image

**Comment by archerzz** [comment]
list of changes:

1. use invariant culture textinfo
2. rename `ToVersionProperty` to `NormalizeVersion`
3. rename variable `s` to `version
4. handle the case when original version starts with `v`, which can cover cases like `Version`, `Ver`, etc
5. add test cases

**Comment by archerzz** [comment]
> This constitutes a breaking change. While I appreciate your analysis and agree that the preview libraries that you've identified are permissible, we cannot move forward with this change without explicit sign-off from the owners of the impacted packages or architect approval.
> 
> Further, this change cannot be applied to existing versions in `Azure.MixedReality.Authentication` as it would be a breaking change to a stable library without an exception from the architects. What is your plan here?

Here is what I plan to do:

- Raise this issue in arch board meeting. We need to get the agreement in arch board.
- Ask individual package owners for the sign-off, including `Azure.MixedReality.Authentication`.

The reason I would introduce a breaking change is because I think in reality the impact should be minimal. Let me take `Azure.MixedReality.Authentication` as an example. Here is the critical part of the codes: https://github.com/Azure/azure-sdk-for-net/blob/b926ce7c1ca60bb5c6f9d3cc52ebf773eae920e7/sdk/mixedreality/Azure.MixedReality.Authentication/src/MixedRealityStsClientOptions.cs#L22-L29
```csharp
        public MixedRealityStsClientOptions(ServiceVersion version = ServiceVersion.V2019_02_28_preview)
        {
            Version = version switch
            {
                ServiceVersion.V2019_02_28_preview => "2019-02-28-preview",
                _ => throw new ArgumentException($"The service version {version} is not supported by this library.", nameof(version))
            };
        }
```

As you can see, the enum is for specifying the API version. However, we found that only one value will be generated for that enum (check [codes here](https://github.com/Azure/azure-sdk-for-net/blob/b926ce7c1ca60bb5c6f9d3cc52ebf773eae920e7/sdk/mixedreality/Azure.MixedReality.Authentication/src/MixedRealityStsClientOptions.cs#L34-L45)) See the discussion in [this issue](https://github.com/Azure/autorest.csharp/issues/2391). That means in reality the enum is very **unlikely** to be referenced, since there is no API version to select.

**Comment by lirenhe** [comment]
Can we have a configuration for autorest to decide the behaivour for ClientOptions.ServiceVersion?
We should by default follow the correct pattern to generate the value but for backward compact, we also allow certain services to keep the existing behavour.

**Comment by archerzz** [comment]
> Can we have a configuration for autorest to decide the behaivour for ClientOptions.ServiceVersion? We should by default follow the correct pattern to generate the value but for backward compact, we also allow certain services to keep the existing behavour.

Yes. That is definitely a solution. However, considering right now there is **only one released SDK** having the case problem, I wonder whether it's worth the efforts.

**Comment by archerzz** [comment]
Per discussion with arch board, I've spilt the work into 3 parts:

- Fix and add backward compatibility in `Azure.MixedReality.Authentication` which is the **only GAed SDK**. Here is the PR: https://github.com/Azure/azure-sdk-for-net/pull/29799
- Fix the case issue in manually written codes. Here is the PR: https://github.com/Azure/azure-sdk-for-net/pull/29808
- For the generated codes, after this PR is merged, we incorporate the change in this PR: https://github.com/Azure/azure-sdk-for-net/pull/29809

**Review by jsquire — APPROVED** [review]
Thanks, @archerzz, for driving [sign-off](https://github.com/Azure/azure-sdk-for-net/pull/29809#issuecomment-1182698634) across the impacted generated and [manually written](https://github.com/Azure/azure-sdk-for-net/pull/29809#issuecomment-1182698634) libraries.   I think we're good to move forward.

**Pull request merged**
